libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1753,35 +1784,95 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 {
 	Type const* leftType = type(_operation.leftExpression());
 	Type const* rightType = type(_operation.rightExpression());
-	TypeResult result = leftType->binaryOperatorResult(_operation.getOperator(), rightType);
-	Type const* commonType = result.get();
-	if (!commonType)
-	{
+	_operation.annotation().isLValue = false;
+	_operation.annotation().isConstant = false;
+
+	// Check if the operator is built-in or user-defined.
+	Result<FunctionDefinition const*> userDefinedOperatorResult = leftType->userDefinedOperator(
+		_operation.getOperator(),
+		*currentDefinitionScope(),
+		false
+	);
```

Comment:
This seems a bit inefficient. For every operator used in the code we start from scratch and scan the whole AST for `using for` directives, then determine which of them provide which operators for the type. I think we would be better off having this information pre-computed and stored as an AST annotation. Or inside the type.

What do you think @chriseth?

By the way, if we agree this is a good idea, this is also something that does not have to be done immediately. To wrap up the PR faster, we could put it in an issue. I think it would actually be simple enough to be `good first issue` for a contributor.

Reply:
I just noticed that we do the same for bound functions so it predates this PR. In that case this should definitely be a separate issue.